### PR TITLE
Add ability to exclude posts and pages from ads

### PR DESCRIPTION
**Pull request by claudiulodro**
Closes https://github.com/Automattic/newspack-issues/issues/115. 

**This PR should be tested together with https://github.com/Automattic/newspack-plugin/pull/482.**

This PR adds a toggle for posts and pages that suppresses ads on the post/page. Currently it's only hooked up to the Newspack Google Ads integration, but we should be able to hook it up to the other supported ad providers if needed.

#### To test:

1. Set up some ads that will display on a post or page: at least one global ad, a widget ad, and an ad block.
Screen Shot 2020-03-11 at 10 54 42 AM

2. In the Document panel of the post, toggle ads off:
Screen Shot 2020-03-11 at 10 54 53 AM

3. Visit post/page again. Observe no ads:
Screen Shot 2020-03-11 at 10 55 08 AM

### CODE REVIEW

File: includes/suppress-ads.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+}
+add_action( 'enqueue_block_editor_assets', 'newspack_enqueue_suppress_ad_assets' );
+
+/**
+ * Get whether ads should be displayed on a screen.
+ *
+ * @param int $post_id Post ID to check (optional, default: current post).
+ * @return bool
+ */
+function newspack_should_show_ads( $post_id = null ) {
```
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
function newspack_ads_should_show_ads( $post_id = null ) {
```

Should probably prefix with the full name of the plugin.

File: includes/suppress-ads.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Register meta for suppressing ads on singles.
+ */
+function newspack_register_suppress_ad_meta() {
+	register_post_meta(
+		'',
+		'newspack_suppress_ads',
```
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
		'newspack_ads_suppress_ads',
```

File: includes/class-newspack-ads-blocks.php [hunk 3]
```diff
@@ -112,6 +112,10 @@ public static function enqueue_view_assets( $type ) {
 	 * Google Ad Manager header code
 	 */
 	public static function insert_google_ad_manager_header_code() {
+		if ( ! newspack_should_show_ads() ) {
```
Comment (jeffersonrabb) [inline includes/class-newspack-ads-blocks.php]:
```suggestion
		if ( ! newspack_ads_should_show_ads() ) {
```

File: includes/class-newspack-ads-blocks.php [hunk 4]
```diff
@@ -131,6 +135,10 @@ public static function insert_google_ad_manager_header_code() {
 	 * Google Ad Manager footer code
 	 */
 	public static function insert_google_ad_manager_footer_code() {
+		if ( ! newspack_should_show_ads() ) {
```
Comment (jeffersonrabb) [inline includes/class-newspack-ads-blocks.php]:
```suggestion
		if ( ! newspack_ads_should_show_ads() ) {
```

File: includes/class-newspack-ads-widget.php [hunk 5]
```diff
@@ -43,6 +43,10 @@ function() {
 	 * @param object $instance The Widget instance.
 	 */
 	public function widget( $args, $instance ) {
+		if ( ! newspack_should_show_ads() ) {
```
Comment (jeffersonrabb) [inline includes/class-newspack-ads-widget.php]:
```suggestion
		if ( ! newspack_ads_should_show_ads() ) {
```

File: includes/suppress-ads.php (hunk 2 above)
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
		'newspack_ads_suppress_ads',
```

File: includes/suppress-ads.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * @package Newspack
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Register meta for suppressing ads on singles.
+ */
+function newspack_register_suppress_ad_meta() {
```
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
function newspack_ads_register_suppress_ad_meta() {
```

File: includes/suppress-ads.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		'',
+		'newspack_suppress_ads',
+		array(
+			'show_in_rest' => true,
+			'single'       => true,
+			'type'         => 'boolean',
+		)
+	);
+}
+add_action( 'init', 'newspack_register_suppress_ad_meta' );
```
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
add_action( 'init', 'newspack_ads_register_suppress_ad_meta' );
```

File: includes/suppress-ads.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Enqueue editor assets for suppressing ads.
+ */
+function newspack_enqueue_suppress_ad_assets() {
```
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
function newspack_ads_enqueue_suppress_ad_assets() {
```

File: includes/suppress-ads.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	if ( 'post' === get_current_screen()->post_type || 'page' === get_current_screen()->post_type ) {
+		wp_enqueue_script( 'newspack-suppress-ads', Newspack_Ads::plugin_url( 'dist/suppress-ads.js' ), [], NEWSPACK_ADS_VERSION, true );
```
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
		wp_enqueue_script( 'newspack-ads-suppress-ads', Newspack_Ads::plugin_url( 'dist/suppress-ads.js' ), [], NEWSPACK_ADS_VERSION, true );
```

File: includes/suppress-ads.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	}
+}
+add_action( 'enqueue_block_editor_assets', 'newspack_enqueue_suppress_ad_assets' );
```
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
add_action( 'enqueue_block_editor_assets', 'newspack_ads_enqueue_suppress_ad_assets' );
```

File: includes/suppress-ads.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ * @param int $post_id Post ID to check (optional, default: current post).
+ * @return bool
+ */
+function newspack_should_show_ads( $post_id = null ) {
+	if ( is_singular() ) {
+		if ( null === $post_id ) {
+			$post_id = get_the_ID();
+		}
+
+		if ( get_post_meta( $post_id, 'newspack_suppress_ads', true ) ) {
```
Comment (jeffersonrabb) [inline includes/suppress-ads.php]:
```suggestion
		if ( get_post_meta( $post_id, 'newspack_ads_suppress_ads', true ) ) {
```

File: src/blocks/ad-unit/view.php [hunk 12]
```diff
@@ -13,6 +13,10 @@
  * @return string Returns the post content with latest posts added.
  */
 function newspack_ads_render_block_ad_unit( $attributes ) {
+	if ( ! newspack_should_show_ads() ) {
```
Comment (jeffersonrabb) [inline src/blocks/ad-unit/view.php]:
```suggestion
	if ( ! newspack_ads_should_show_ads() ) {
```

File: src/suppress-ads/index.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+					onChange={ value => {
+						updateSuppressAds( value, meta );
+					} }
+				/>
+			</PluginDocumentSettingPanel>
+		);
+	}
+};
+
+const ComposedPanel = compose( [
```
[the remark this answers is not in the document]
Comment (claudiulodro) [reply src/suppress-ads/index.js]: Much nicer! Thanks for the suggestion; I've updated at https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-4fcb1c4b288066f0a515a7484e1a4c56R39-R45

File: src/suppress-ads/index.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * WordPress dependencies
+ */
+import { registerPlugin } from '@wordpress/plugins';
+import { PluginDocumentSettingPanel } from '@wordpress/edit-post';
+import { __ } from '@wordpress/i18n';
+import { ToggleControl } from '@wordpress/components';
+import { Component } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { withDispatch, withSelect, select } from '@wordpress/data';
```
[the remark this answers is not in the document]
Comment (claudiulodro) [reply src/suppress-ads/index.js]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-4fcb1c4b288066f0a515a7484e1a4c56R12

File: src/suppress-ads/index.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+ * Add a section to the Document settings with a toggle for suppressing ads on the current single.
+ */
+class NewspackSuppressAdsPanel extends Component {
+	render() {
+		const { meta, updateSuppressAds } = this.props;
+
+		return (
+			<PluginDocumentSettingPanel
+				name="newspack-ad-free"
+				title={ __( 'Suppress Ads', 'newspack' ) }
```
[the remark this answers is not in the document]
Comment (claudiulodro) [reply src/suppress-ads/index.js]: Great idea. https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-4fcb1c4b288066f0a515a7484e1a4c56R23

File: src/suppress-ads/index.js [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+				className="newspack-subtitle"
+			>
+				<ToggleControl 
+					label={ __( 'Don\'t show ads on this post or page', 'newspack' ) }
```
[the remark this answers is not in the document]
Comment (claudiulodro) [reply src/suppress-ads/index.js]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-4fcb1c4b288066f0a515a7484e1a4c56R27

File: includes/suppress-ads.php (hunk 1 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: That makes sense. Good suggestion. https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R13

File: includes/suppress-ads.php (hunk 2 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R16

File: includes/class-newspack-ads-blocks.php (hunk 3 above)
Comment (claudiulodro) [reply includes/class-newspack-ads-blocks.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-f00ee91a5e4c5891df399ba5eae186dcR115

File: includes/class-newspack-ads-blocks.php (hunk 4 above)
Comment (claudiulodro) [reply includes/class-newspack-ads-blocks.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-f00ee91a5e4c5891df399ba5eae186dcR138

File: includes/class-newspack-ads-widget.php (hunk 5 above)
Comment (claudiulodro) [reply includes/class-newspack-ads-widget.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-437dce7992d283d1e00b32ca115aab2eR46

File: includes/suppress-ads.php (hunk 2 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R16

File: includes/suppress-ads.php (hunk 6 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R13

File: includes/suppress-ads.php (hunk 7 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R24

File: includes/suppress-ads.php (hunk 8 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R29

File: includes/suppress-ads.php (hunk 9 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R31

File: includes/suppress-ads.php (hunk 10 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R34

File: includes/suppress-ads.php (hunk 11 above)
Comment (claudiulodro) [reply includes/suppress-ads.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-7af73f10e634570be7ed9d3b5db857d2R48

File: src/blocks/ad-unit/view.php (hunk 12 above)
Comment (claudiulodro) [reply src/blocks/ad-unit/view.php]: https://github.com/Automattic/newspack-ads/pull/26/commits/103be8c309059f4806d9a4b4ac5f5d080649bfef#diff-d8c625c059bed804156ed322fe811755R16

**Comment by claudiulodro** [comment]
Thanks for the review! I've updated this PR with your recommended changes.

**Pull request merged**